internal/mining/mining.go

```diff
@@ -922,6 +930,132 @@ func (g *BlkTmplGenerator) handleTooFewVoters(nextHeight int64, miningAddress st
+	}
+	ticketUtxo, err := g.cfg.FetchUtxoEntry(ticketSubmission)
+	if err != nil {
+		return nil, makeError(ErrGetTicketInfo, err.Error())
+	}
+	if ticketUtxo == nil || ticketUtxo.IsSpent() {
+		str := fmt.Sprintf("ticket %v does not exist or is spent", ticketHash)
+		return nil, makeError(ErrGetTicketInfo, str)
+	}
+
+	// Add the ticket submission utxo to the block utxos view so that it is
+	// available for lookup later.
+	blockUtxos.Entries()[ticketSubmission] = ticketUtxo
+
+	// Get the minimal outputs for the ticket.
+	ticketMinOuts := ticketUtxo.TicketMinimalOutputs()
+	if ticketMinOuts == nil {
+		str := fmt.Sprintf("ticket %v missing minimal outputs", ticketHash)
+		return nil, makeError(ErrGetTicketInfo, str)
+	}
+
+	// Create a revocation transaction for the ticket.
+	const isAutoRevocationsEnabled = true
+	const revocationTxFee = dcrutil.Amount(0)
+	revocationMsgTx, err := stake.CreateRevocationFromTicket(ticketHash,
+		ticketMinOuts, revocationTxFee, stake.TxVersionAutoRevocations,
+		g.cfg.ChainParams, prevHeaderBytes, isAutoRevocationsEnabled)
+	if err != nil {
+		return nil, err
+	}
+	revocationTx := dcrutil.NewTx(revocationMsgTx)
+	revocationTx.SetTree(wire.TxTreeStake)
+	txDesc := &TxDesc{
+		Tx:   revocationTx,
+		Type: stake.TxTypeSSRtx,
+		TotalSigOps: g.cfg.CountSigOps(revocationTx, false, false,
+			isTreasuryEnabled),
+		TxSize: int64(revocationMsgTx.SerializeSize()),
+	}
+
+	return txDesc, nil
+}
+
+// addAutoRevocationsToQueue is a helper function that creates revocations for
+// all tickets that will become missed or expired as of the block being created.
+// It inserts the created revocations into the provided priority queue and maps.
+//
+// Note: This should only be used if the automatic ticket revocations agenda is
+// active.
+func (g *BlkTmplGenerator) addAutoRevocationsToQueue(winningTickets map[chainhash.Hash]bool,
+	blockUtxos *blockchain.UtxoViewpoint, prevHeaderBytes []byte, numSSGen int,
+	isTreasuryEnabled bool, priorityQueue *txPriorityQueue,
+	prioritizedTxns map[chainhash.Hash]struct{},
+	prioItemMap map[chainhash.Hash]*txPrioItem) error {
+
+	// Create a slice of the ticket hashes that must be revoked due to
+	// becoming missed or expired this block.
+	best := g.cfg.BestSnapshot()
+	missedCount := len(best.NextWinningTickets) - numSSGen
+	expiredCount := len(best.NextExpiringTickets)
+	revokeTickets := make([]chainhash.Hash, 0, missedCount+expiredCount)
+
+	// Add tickets that will be missed to the slice of ticket hashes to
+	// revoke.
+	for ticketHash, hasVote := range winningTickets {
+		// If a winning ticket does not have a vote in this block, it will be
+		// missed.
+		if !hasVote {
+			revokeTickets = append(revokeTickets, ticketHash)
+		}
+	}
+
+	// Add tickets that will be expired to the slice of ticket hashes to
+	// revoke.
+	revokeTickets = append(revokeTickets, best.NextExpiringTickets...)
+
+	// Return now if there are no tickets to revoke.
+	if len(revokeTickets) == 0 {
```

Comment:
Nit: this check could be moved up and changed to `expiredCount+missedCount == 0` to avoid the append allocations for `revokeTickets`.

Reply:
Good point, no reason to continue if  `expiredCount+missedCount == 0`.  Updated.